Pause jobs with invalid inputs

mvdbeek opened the pull request:
Otherwise these jobs may hang around forever/until the admin
does a cleanup. This is probably preferable for users as well,
as the pause message gives a hint of why their job didn't start
running. Fixes #6022
and would minimize maintenance as outlined in #5944.

I'm working on tests, but one easy way to test this is create a collection
through the list builder interface, then delete the source HDAs.
Then run a workflow on the collection and without this commit
the outputs get created but the job just hangs around indefinetely.
With the PR the steps consuming the collection and the subsequent jobs
will be paused.

This should also help with deleting histories for running workflows and other edge cases.

mvdbeek committed:
Pause jobs with invalid inputs

Otherwise these jobs may hang around forever/until the admin
does a cleanup. This is probably preferable for users as well,
as the pause message gives a hint of why their job didn't start
running. Fixes https://github.com/galaxyproject/galaxy/issues/6022
and would minimize maintenance as outlined in https://github.com/galaxyproject/galaxy/issues/5944.

I'm working on tests, but one easy way to test this is create a collection
through the list builder interface, then delete the source HDAs.
Then run a workflow on the collection and without this commit
the outputs get created but the job just hangs around indefinetely.
With the PR the steps consuming the collection and the subsequent jobs
will be paused.

Set to paused

mvdbeek committed:
Add a testcase for pausing jobs with deleted input

Fix test

natefoo commented:
Do we want to fail jobs w/ inputs in `failed_metadata` state? While they can't be fixed currently, this is due to a bug - it should be possible to run auto-detect on a dataset whose inputs are not yet queued. This can't be done currently because there's no differentiation between "queued in Galaxy", "dispatched to the runner plugin" and "queued in the DRM", but [I'm working on that](https://github.com/natefoo/galaxy/tree/job-feedback).

Also if an input is deleted but not purged, you could undelete the dataset to allow it to run, we just don't provide any feedback to users indicating what the problem is.

mvdbeek commented:
I agree with those things, but it's easy enough to correct this once the job fails. I.e if you have a workflow or you've queued up jobs the first job fails and the rest is paused, so you know you can take an action and resume from there on.

> Also if an input is deleted but not purged, you could undelete the dataset to allow it to run, we just don't provide any feedback to users indicating what the problem is.

That's why I'm recording the reason for the failure and write it into the info field, so a user can take action. If there's jobs depending on that job that has a deleted input they will be resumed if the users clicks `resume dependencies`

mvdbeek commented:
Skimming through your code I think this would co-exist nicely, since those additional states are non-terminal, right ?

natefoo commented:
Okay, so here's a scenario on a fresh Galaxy instance, with a 4 step workflow where each tool has 1 input and 1 output (except the first tool is a data source tool of some kind, so no input):

1. Job 1's output HDA ends in state `failed_metadata`
2. Job 2 and its output HDA/Dataset are set to state `error`
3. Jobs 3 & 4 and their output HDAs/Datasets are set to state `paused`
4. Use metadata auto-detect on HDA 1
5. Use rerun remap on HDA 2 to replace error'd HDA/Dataset
6. Unpause 3 & 4

The same process would be used if you deleted HDA/Job 1 while it was running? Do I have this right?

I think my confusion came from failing Job/HDA 2, my assumption was that we'd pause that one and once Job/HDA 1 was fixed, you could simply unpause 2. Step 5 shouldn't be necessary.

Yes, those states are non-terminal so this shouldn't be an issue - my previous comment about a lack of differentiation in queued states is irrelevant I suppose because those jobs would be moved to `paused` or `error` (and they're currently `new` anyway, not `queued`).

mvdbeek commented:
> The same process would be used if you deleted HDA/Job 1 while it was running? Do I have this right?

Almost, the first scenario is correct, the second scenario I think would (and is already) be handled by calling `job_wrapper.fail` outside of this query (gotta dig where exactly if you want to know, probably during finishing of the job), but the net effect is the same.

> I think my confusion came from failing Job/HDA 2, my assumption was that we'd pause that one and once Job/HDA 1 was fixed, you could simply unpause 2. Step 5 shouldn't be necessary.

I can change this to paused, that is indeed the better option.

natefoo commented:
This sounds great!

mvdbeek commented:
Alright @natefoo, now we just pause everything with invalid inputs since we should be able to resume once the issue has been resolved.

mvdbeek commented:
Meh, that's a legitimate failure, but not quite sure yet what the issue is.

mvdbeek committed:
Fail jobs with purged inputs

We can't resume these jobs anyway, so better to set them to failed.
Also resets the info field for datasets produced by resumed jobs.

mvdbeek committed:
Display the info field for paused jobs

mvdbeek committed:
Pause job if input dataset state is paused

mvdbeek commented:
I've also changed the way we display paused datasets to the user. We now include the reason for a dataset being paused (if there is any).
<img width="225" alt="screen shot 2018-05-06 at 11 29 40" src="https://user-images.githubusercontent.com/6804901/39674688-c2b27134-514f-11e8-8c1f-6f5c9adf4c6a.png">

natefoo commented:
Is this ready for review?

mvdbeek commented:
It is!

mvdbeek committed:
Optimize job resume function

natefoo commented on client/galaxy/scripts/mvc/dataset/dataset-li.js:
```diff
@@ -502,7 +502,8 @@ DatasetListItemView.prototype.templates = (() => {
     summaryTemplates[STATES.PAUSED] = BASE_MVC.wrapTemplate(
-        ["<div>", _l('This job is paused. Use the "Resume Paused Jobs" in the history menu to resume'), "</div>"],
+        ["<div>", _l('This job is paused. Use the "Resume Paused Jobs" in the history menu to resume'), "</div>",
+        '<div><%- dataset.misc_info %></div>'],
```
`<div class="info">` puts the info in to a box so it has a bit of separation from the resume message.

mvdbeek replied:
Ooh, that does look nicer
![screen shot 2018-05-31 at 16 10 10](https://user-images.githubusercontent.com/6804901/40787245-6fe69974-64ed-11e8-9b01-ce9719fc5825.png)

natefoo reviewed (approved):
This works great, I only have one concern - the jobs-ready-to-run query is the behemoth that it is for performance reasons. Moving the input state checking to the code and creating a new query per waiting job on each loop could be a major performance problem. While there should be a drastic decrease in `NEW` state jobs that would mitigate the impact, we should probably take a copy of a database with a lot of `NEW` state jobs (e.g. usegalaxy.org's) and run this on it to see how it performs on the first start, and thereafter.

mvdbeek committed:
Use info style when displaying the reason for pausing a dataset

mvdbeek commented:
Yeah, I guess that the calls to pause and fail jobs may take a bit of time if you have a lot of jobs in these states, since there is some flushing involved per job. We could optimize this some more, but on the other hand I'm not sure this is really an issue and we're not optimizing prematurely.

natefoo commented:
I started this against a copy of test.galaxyproject.org's database and it took it from:

```
    state    |   ct   
-------------+--------
 paused      |   3458
 new         |   2423
```

To:

```
  state  |   ct   
---------+--------
 paused  |   4827
 new     |     58
```

(Plus errors due to missing tools in my test/dev instance). The remaining new state jobs are partially due to at least one input being non-terminal, although there's a few oddballs:

```sql
test_test=> select j.id, j.state, hda.id, hda.name, hda.state, hda.deleted, hda.purged, d.id, d.state, d.deleted, d.purged from job j left outer join job_to_input_dataset jtid on j.id = jtid.job_id left outer join history_dataset_association hda on hda.id = jtid.dataset_id left outer join dataset d on hda.dataset_id = d.id where j.id = 237474 order by j.id desc;
   id   | state |   id   |                                 name                                  | state | deleted | purged |   id   | state | deleted | purged 
--------+-------+--------+-----------------------------------------------------------------------+-------+---------+--------+--------+-------+---------+--------
 237474 | new   | 432828 | Get Microbial Data (CDS for Acaryochloris marina MBIC11017:NC_009929) |       | f       | f      | 397850 | error | f       | f
(1 row)
```

I'm not sure why this one isn't being dealt with, but overall this is a vast improvement.

mvdbeek commented:
I think this example goes straight to ignore in https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/jobs/handler.py#L410 because the dataset state is error, which should normally be handled by whatever made the input fail ? It's also a data source tool, they're a bit special. Maybe we want to fail those as well ?

natefoo commented:
Ah so the issue is likely that most (all?) of these jobs were created many years ago before a dataset ending in error would pause any dependent jobs/datasets.

mvdbeek commented:
Seems plausible, can you check that ?

natefoo commented:
Yeah, ok, most are from 2013 or earlier. Of the remaining, causes include

- User is not active
- input Dataset state is `error` or `queued`
- input HDA id is `null`

Most of these are probably caused by unsafe server shutdowns. For inactive users, we should probably just refuse to create jobs/datasets/HDAs until they are active.

natefoo commented:
One final update: After manually erroring all jobs of the previous type plus:

- input HDA or Dataset is deleted
- Job.handler is invalid (changed handler names)

I now have:

```sql
galaxy_test=> select count(id) from job where state = 'new';
 count 
-------
     0
(1 row)
```

Of course, I'll accumulate some new ones over time, but there shouldn't be very many. Thanks again @mvdbeek!

nsoranzo commented:
It seems my Galaxy server had something like 60,000 jobs in `new` state, after updating to 18.09 my handlers have been blocked for hours doing just:
```
galaxy.jobs DEBUG 2018-10-18 22:40:36,900 Pausing Job '995223', Execution of this dataset's job is paused because its input datasets are in an error state.
galaxy.tools.error_reports DEBUG 2018-10-18 22:40:37,692 Bug report plugin <galaxy.tools.error_reports.plugins.sentry.SentryPlugin object at 0x7f6cb977bf10> generated response None
galaxy.jobs DEBUG 2018-10-18 22:40:38,012 Pausing Job '995227', Execution of this dataset's job is paused because its input datasets are in an error state.
galaxy.tools.error_reports DEBUG 2018-10-18 22:40:38,608 Bug report plugin <galaxy.tools.error_reports.plugins.sentry.SentryPlugin object at 0x7f6cb977bf10> generated response None
...
```
Is this something we should mention in the release notes?

natefoo commented:
> Is this something we should mention in the release notes?

Yes, that seems like a good idea. Would you like me to write something?

nsoranzo commented:
@natefoo That would be much appreciated!

By the way, I still have 7405 job in `new` state after this, any suggestion?

mvdbeek commented:
@nsoranzo could it be what nate described in https://github.com/galaxyproject/galaxy/pull/6036#issuecomment-393900631 ?
Are those jobs very old ?

natefoo commented:
I added a warning to admins in 3263aba44a95e2a5e50e5050893628ba618fd306.

nsoranzo commented:
> could it be what nate described in #6036 (comment) ?

@mvdbeek After a small investigation, all these jobs have at least one input dataset for each of the 3 `new`, `error` and `ok` states:
```
prod_galaxy=> SELECT COUNT(DISTINCT job.id) FROM job JOIN job_to_input_dataset jtid ON job.id = jtid.job_id JOIN history_dataset_association hda ON hda.id = jtid.dataset_id JOIN dataset d ON hda.dataset_id = d.id WHERE job.state = 'new' HAVING bool_or(d.state = 'new');
 count 
-------
  7405
(1 row)

prod_galaxy=> SELECT COUNT(DISTINCT job.id) FROM job JOIN job_to_input_dataset jtid ON job.id = jtid.job_id JOIN history_dataset_association hda ON hda.id = jtid.dataset_id JOIN dataset d ON hda.dataset_id = d.id WHERE job.state = 'new' HAVING bool_or(d.state = 'paused');
 count 
-------
(0 rows)

prod_galaxy=> SELECT COUNT(DISTINCT job.id) FROM job JOIN job_to_input_dataset jtid ON job.id = jtid.job_id JOIN history_dataset_association hda ON hda.id = jtid.dataset_id JOIN dataset d ON hda.dataset_id = d.id WHERE job.state = 'new' HAVING bool_or(d.state = 'error');
 count 
-------
  7405
(1 row)

prod_galaxy=> SELECT COUNT(DISTINCT job.id) FROM job JOIN job_to_input_dataset jtid ON job.id = jtid.job_id JOIN history_dataset_association hda ON hda.id = jtid.dataset_id JOIN dataset d ON hda.dataset_id = d.id WHERE job.state = 'new' HAVING bool_or(d.state = 'ok');
 count 
-------
  7405
(1 row)
```

> Are those jobs very old ?

Not really, oldest is from 2016-06-17, most recent is from 3 days ago.

> I added a warning to admins in 3263aba.

Thanks @natefoo!

mvdbeek commented:
> at least one input dataset for each of the 3 `new`, `error` and `ok` states:

That's curious, so all of them should have been marked as paused because of the HDA state being error. What I'm doing in the code is checking if the Dataset state is not OK -- any chance the dataset state for the HDA is actually OK and only the HDA state is error ? (I'm also checking for failed metadata, so if that's the case I'm curious what happened there)

nsoranzo commented:
@mvdbeek I think the issue is that, since some input dataset is in `new` state, these jobs are not selected when first calculating `jobs_to_check` (due to `~model.Job.table.c.id.in_(hda_not_ready)`), and therefore are never processed by `__filter_jobs_with_invalid_input_states()` .

mvdbeek commented:
Right, that makes sense!

natefoo commented:
What'd be a good solution for this? Some sort of secondary loop that finds such jobs? I worry that moving dataset state checking from the DB query to the code would have a negative performance impact.

@nsoranzo any insight as to why the jobs' other inputs are not eventually moving to a terminal state?

mvdbeek commented:
To prevent this I guess we should set all jobs that depend on a failed dataset to paused in the job finish context, once that works we just have to do a one-time cleanup. I think we already do this, but apparently we miss some cases.

natefoo commented:
Ah yes that makes a lot more sense.

nsoranzo commented:
Actually I was wrong with my SQL queries (forgot `GROUP BY job.id`), there are only 2 jobs with input datasets in 'new', 'error' and 'ok' states. Instead:
- 3546 jobs have all input datasets in either 'new' or 'ok' (1299 all 'new');
- 3857 jobs have all input datasets in either 'error' or 'ok' (1049 all 'error').

The 1049 jobs with all 'error' input datasets are particularly surprising.
